Add EIP: Social Recovery Interface (#7093)
ethereum/EIPs: EIPS/eip-7093.md

---
eip: 7093
title: Social Recovery Interface
description: Interfaces for social recovery account supporting various guardian types and customizable recovery policies.
author: John Zhang (@johnz1019), Davis Xiang (@xcshuan), Kyle Xu (@kylexyxu), George Zhang (@odysseus0)
discussions-to: https://ethereum-magicians.org/t/eip-social-recovery-interface/14494
status: Draft
type: Standards Track
category: ERC
created: 2023-05-29

> ---

Reviewer: If you recommend that implementations support EIP-1271 signatures, you should require that EIP here:

```suggestion
requires: 1271
---
```

> 
> 

Reviewer: ```suggestion
```

## Abstract

> This ERC proposes a standard interface for social recovery of smart contract accounts. It separates identity and policy verification from the recovery process, allowing more types of Guardians than just on-chain accounts. It also lets users customize recovery policies without changing the account’s smart contract.

Reviewer: It'd be nice to expand the definition of "Guardian" here in the abstract, maybe something like:

```suggestion
This ERC proposes a standard interface for social recovery of smart contract accounts. It separates identity and policy verification from the recovery process, allowing more ways to authenticate (known as Guardians) than just on-chain accounts. It also lets users customize recovery policies without changing the account’s smart contract.
```

1. Appoint friends or family members, who do not have blockchain accounts, as Guardians for social recovery.
2. Use NFTs/SBTs as Guardians for their accounts.
3. Personalize and implement adaptable recovery policies.
4. Support novel types of Guardians and recovery policies without needing to upgrade their account contracts.
5. Enable multiple recovery mechanism support, thereby eliminating single points of failure.

This approach enables users to customize recovery policies without the need to change the smart contract of the account itself.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

This EIP consists of four key concepts:

> - **Identity**: This denotes the representation of a Guardian's identity on the blockchain. It encapsulates traditional on-chain account types such as Externally Owned Accounts (EOA) and Smart Contract Accounts (SCA). More importantly, it extends to include any identity construct capable of producing construct able to be verified on-chain, like signatures and proofs. This could range from Webauthn/Passkey R1 keys to Email DomainKeys Identified Mail (DKIM) signatures, OpenID tokens, Zero-Knowledge Proofs (ZKP), Non-Fungible Tokens (NFTs), SoulBound Tokens (SBTs), and even types yet to be developed. This comprehensive approach ensures a broad, forward-compatible support for various identity types.

Reviewer: We (the EIP editors) have a call later this week to confirm that we want to allow links to w3c. Assuming that goes through, I'd like to see a link to the webauthn standard here (but that can wait until after we merge this PR.)

> - **GuardianPermissionVerifier**: This component defines how to verify the signature or proof provided by the guardian. Regardless of whether the Guardian's account is on-chain or off-chain, the GuardianPermissionVerifier is invoked during the recovery process of smart contract accounts that incorporate a social recovery system. Its primary role is to confirm the validity of the Guardian's signature or proof, thereby ensuring the authenticity of the Guardian during the recovery process.

Reviewer: ```suggestion
- **GuardianPermissionVerifier**: This component defines how to verify the signature or proof provided by the Guardian. Regardless of whether the Guardian's account is on-chain or off-chain, the GuardianPermissionVerifier is invoked during the recovery process of smart contract accounts that incorporate a social recovery system. Its primary role is to confirm the validity of the Guardian's signature or proof, thereby ensuring the authenticity of the Guardian during the recovery process.
```

- **RecoveryPolicyVerifier**: This component offers a flexible interface for validating recovery policies. The flexibility stems from allowing account holders or authorized parties to define and store their recovery policies. During the recovery process, the verification logic is implemented by invoking the specific function of the contract instance adopting this interface. Thus, a wide array of customizable social recovery scenarios can be catered to through different contract instances and policy configurations.
- **RecoveryAccount**: This component encapsulates the core of the social recovery functionality. It is designed to be flexible, composable, and extensible to adapt to various recovery needs. Each RecoveryAccount is defined by an instance contract, crafted by smart contract developers, which embeds the essential logic for the recovery process.

![social_recovery_flow](../assets/eip-7093/social-recovery-flow.png)

### DataTypes

> ### `TypesAndDecoders.sol`

Reviewer: I would strongly recommend against organizing your EIP by file name. Since there will likely be multiple implementations of this standard, they will all organize their code differently. How code is organized is _not_ part of a good EIP, since it doesn't affect the observable properties of an implementation.

This defines the necessary data types required by this interface standard.

```solidity
pragma solidity ^0.8.13;

> // SPDX-License-Identifier: MIT

Reviewer: Solidity within the body of the EIP _must_ be `CC0-1.0` licensed.

    address guardianVerifier;
    bytes signer;
}

/**
 * @dev Structure representing a guardian with a property
 * The property of Guardian are defined by the associated RecoveryPolicyVerifier contract.
 */
struct GuardianInfo {
    Identity guardian;
    uint32 property; //eg.,Weight,Percentage,Role with weight,etc.
}

/**
 * @dev Structure representing a threshold configuration
 */
struct ThresholdConfig {
    uint32 threshold; // Threshold value
    int48 lockPeriod; // Lock period for the threshold
}

/**
 * @dev Structure representing a recovery configuration
 * A RecoveryConfig can have multiple threshold configurations for different threshold values and their lock periods
 */
struct RecoveryConfig {
    GuardianInfo[] guardianInfos;
    ThresholdConfig[] thresholdConfigs;
}

/**
 * @dev Structure representing a recovery policy with its verification logic (policy verifier contract address)
 */
struct RecoveryPolicy {
    address policyVerifier;
    RecoveryConfig config;
}

struct Permission {
    Identity guardian;
    bytes signature;
}

```

The `Identity` structure represents various types of guardians. The process of identity verification is as follows:

> - When the `signer` value in the declared entity is empty, this implies that the `Identity` entity is of EOA/SCA account type. In this case, `guardianVerifier` address should be the address of EOA/SCA (the actual signer). For permission verification of this `Identity` entity, one can use `SignatureChecker.isValidSignatureNow` from the Openzeppelin library, which supports both ECDSA and [ERC-1271](./eip-1271.md) signature verifications.

Reviewer: Don't make recommendations on implementation. Instead describe the properties you'd like the implementation to have. For example, you may want to recommend implementers watch out for signature malleability attacks (which is what is provided by OZ as far as I know.)

     */
    function isGuardian(identity guardian) public view returns (bool, RecoveryPolicy[] recoveryPolicyConfigs);

    function getRecoveryPolicies() public view returns (RecoveryPolicy[] recoveryPolicyConfigs);

    function getRecoveryConfigs(address policyVerifier) public view returns (RecoveryConfig config);

    function getRecoveryNonce() public view returns (uint256 nonce);

    function getRecoveryStatus(address policyVerifier) public view returns (bool isRecovering, uint48 expiryTime);
}

```

- For the `Guardian`'s signable message, it SHOULD employ [EIP-712](./eip-712.md) type signature to ensure the content of the signature is readable and can be confirmed accurately during the Guardian signing process.
- `getRecoveryNonce()` SHOULD be separated from nonces associated with account asset operations, as social recovery is a function at the account layer.

### **Recovery Account Workflow**

Note: This workflow is presented as an illustrative example to clarify the coordinated usage of the associated interface components. It does not imply a mandatory adherence to this exact process.

1. A user sets up a `recoveryPolicyConfigA` within his `RecoveryAccount`:

>    ```json

Reviewer: This blob doesn't look like valid JSON.

   - Verify that `guardians` (Identity A and B) are present in `config.guardianInfos` list and are unique.
   - Accumulate the `property` values of `guardians` (30 + 30 = 60).
   - Compare the calculated result (60) with the `config.thresholdConfigs.threshold` ,the result is more than the first element (`threshold: 50, lockPeriod: 24 hours`) but less than the second element (`threshold: 100, lockPeriod: ""`), the validation is successful, and the lock period of 24 hours is returned.

6. The `RecoveryAccount` saves a temporary state `{policyVerifier, newOwners, block.timestamp + 24 hours}` and increments `RecoveryNonce`. A `RecoveryStarted` event is emitted.

7. After the expiry time, anyone (usually a relayer) can call `RecoveryAccount.executeRecovery()` to replace `newOwners`, remove the temporary state, complete the recovery, and emit a `RecoveryExecuteed` event.

   

## Rationale

This standard aims to define interfaces for social recovery in smart contract accounts, enabling diverse Guardian types and flexible, customizable recovery policies. The goal is to ensure compatibility and security through a standardized recovery process.

> By abstracting identity verification and recovery policy verification interfaces, we decouple the verification logic from the social recovery process. This design allows the seamless addition of new identity verification methods or changes in recovery policies without altering the core recovery logic.

Reviewer: This seems more like content for the motivation section. The rationale section should justify individual technical choices made within the EIP itself, while the motivation should justify the EIP as a whole. For example, you might want to explain why you use a recovery nonce in this section.

1. Appoint friends or family members, who do not have blockchain accounts, as Guardians for social recovery.
2. Use NFTs/SBTs as Guardians for their accounts.
3. Personalize and implement adaptable recovery policies.
4. Support novel types of Guardians and recovery policies without needing to upgrade their account contracts.
5. Enable multiple recovery mechanism support, thereby eliminating single points of failure.

This approach enables users to customize recovery policies without the need to change the smart contract of the account itself.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

This EIP consists of four key concepts:

> - **Identity**: This denotes the representation of a Guardian's identity on the blockchain. It encapsulates traditional on-chain account types such as Externally Owned Accounts (EOA) and Smart Contract Accounts (SCA). More importantly, it extends to include any identity construct capable of producing construct able to be verified on-chain, like signatures and proofs. This could range from [Webauthn](https://www.w3.org/TR/webauthn-2/)/Passkey R1 keys to Email DomainKeys Identified Mail (DKIM) signatures [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376), OpenID tokens, Zero-Knowledge Proofs (ZKP), Non-Fungible Tokens (NFTs), SoulBound Tokens (SBTs), and even types yet to be developed. This comprehensive approach ensures a broad, forward-compatible support for various identity types.

Reviewer: You have to specify an exact version of the w3 specification:

```suggestion
- **Identity**: This denotes the representation of a Guardian's identity on the blockchain. It encapsulates traditional on-chain account types such as Externally Owned Accounts (EOA) and Smart Contract Accounts (SCA). More importantly, it extends to include any identity construct capable of producing construct able to be verified on-chain, like signatures and proofs. This could range from [Webauthn](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/)/Passkey R1 keys to Email DomainKeys Identified Mail (DKIM) signatures [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376), OpenID tokens, Zero-Knowledge Proofs (ZKP), Non-Fungible Tokens (NFTs), SoulBound Tokens (SBTs), and even types yet to be developed. This comprehensive approach ensures a broad, forward-compatible support for various identity types.
```
